File: core/src/api/properties/PropertyStore.ts
import { Property } from "./Property";
import { PropertyValues } from "./PropertyValues";

/**
 * A store that contains properties indexed by their property key.
 *
 * @typeParam VALUE - The type of property values, must extend PropertyValues
 * @typeParam PROPERTY - The type of property, must extend Property<VALUE>
 */
export interface PropertyStore<
  VALUE extends PropertyValues,
  PROPERTY extends Property<VALUE>
> {
  /**
   * Returns a map of property keys to properties.
   *
   * @returns Map of property keys to properties
   */
  properties(): Map<string, PROPERTY>;
}

/**
 * Provides default implementations and utilities for PropertyStore interface.
 */
export namespace PropertyStore {
  /**
   * Returns a map of property keys to property values.
   *
   * @param store The property store
   * @returns Map of property keys to property values
   */
  export function propertyValues<
    VALUE extends PropertyValues,
    PROPERTY extends Property<VALUE>
  >(store: PropertyStore<VALUE, PROPERTY>): Map<string, VALUE> {
    const result = new Map<string, VALUE>();
    store.properties().forEach((property, key) => {
      result.set(key, property.values());
    });
    return result;
  }

  /**
   * Gets a property by key.
   *
   * @param store The property store
   * @param propertyKey The property key
   * @returns The property, or undefined if not found
   */
  export function get<
    VALUE extends PropertyValues,
    PROPERTY extends Property<VALUE>
  >(
    store: PropertyStore<VALUE, PROPERTY>,
    propertyKey: string
  ): PROPERTY | undefined {
    return store.properties().get(propertyKey);
  }

  /**
   * Checks if the property store is empty.
   *
   * @param store The property store
   * @returns True if the store contains no properties
   */
  export function isEmpty<
    VALUE extends PropertyValues,
    PROPERTY extends Property<VALUE>
  >(store: PropertyStore<VALUE, PROPERTY>): boolean {
    return store.properties().size === 0;
  }

  /**
   * Returns the set of property keys in this store.
   * Derived value - will return an immutable set.
   *
   * @param store The property store
   * @returns Set of property keys
   */
  export function keySet<
    VALUE extends PropertyValues,
    PROPERTY extends Property<VALUE>
  >(store: PropertyStore<VALUE, PROPERTY>): ReadonlySet<string> {
    return new Set(store.properties().keys());
  }

  /**
   * Checks if the property store contains a property with the given key.
   *
   * @param store The property store
   * @param propertyKey The property key to check
   * @returns True if the property exists
   */
  export function containsKey<
    VALUE extends PropertyValues,
    PROPERTY extends Property<VALUE>
  >(store: PropertyStore<VALUE, PROPERTY>, propertyKey: string): boolean {
    return store.properties().has(propertyKey);
  }
}

File: gds/src/types/properties/property_store.rs
use super::property::Property;
use crate::types::ValueType;
use thiserror::Error;

/// Error type for property store operations.
#[derive(Error, Debug, Clone, PartialEq, Eq)]
pub enum PropertyStoreError {
    #[error("Property not found: {0}")]
    PropertyNotFound(String),

    #[error("Property already exists: {0}")]
    PropertyAlreadyExists(String),

    #[error("Invalid property key: {0}")]
    InvalidPropertyKey(String),

    #[error("Property '{key}' declares {declared:?} but its values are {materialized:?}")]
    SchemaValueTypeMismatch {
        key: String,
        declared: ValueType,
        materialized: ValueType,
    },
}

pub type PropertyStoreResult<T> = Result<T, PropertyStoreError>;

pub(crate) fn validate_column_value_type(
    key: &str,
    declared: ValueType,
    materialized: ValueType,
) -> PropertyStoreResult<()> {
    if declared == materialized {
        return Ok(());
    }
    Err(PropertyStoreError::SchemaValueTypeMismatch {
        key: key.to_string(),
        declared,
        materialized,
    })
}

/// A framing protocol for schema-bearing property columns.
pub trait PropertyStore: Send + Sync {
    type Property: Property;

    /// Gets a column by its schema key.
    fn get(&self, property_key: &str) -> Option<&Self::Property>;

    /// Iterates over the columns without exposing their storage representation.
    fn columns(&self) -> Box<dyn Iterator<Item = &Self::Property> + '_>;

    /// Adds a new column, rejecting duplicate schema keys.
    fn add_column(&mut self, property: Self::Property) -> PropertyStoreResult<()>;

    /// Replaces an existing column with the same schema key.
    fn replace_column(&mut self, property: Self::Property) -> PropertyStoreResult<Self::Property>;

    /// Removes and returns a column by schema key.
    fn remove_column(&mut self, property_key: &str) -> PropertyStoreResult<Self::Property>;

    /// Checks if the property store is empty.
    fn is_empty(&self) -> bool {
        self.len() == 0
    }

    /// Returns the number of properties in the store.
    fn len(&self) -> usize {
        self.columns().count()
    }

    /// Returns the set of property keys in this store.
    fn key_set(&self) -> Vec<&str> {
        self.columns()
            .map(|property| property.schema().key())
            .collect()
    }

    /// Checks if the store contains a property with the given key.
    fn contains_key(&self, property_key: &str) -> bool {
        self.get(property_key).is_some()
    }
}
